### ISSUE TITLE
8.0.1578 test failure on macOS

### ISSUE THREAD

**Issue by chdiza** [vim/vim#2693]
On macOS 10.3.3, 8.0.1578 has test failures that don't seem to have been caught by CI:
```
From test_alot.vim:
Found errors in Test_popup_command():
function RunTheTest[38]..Test_popup_command[13]..VerifyScreenDump line 14: See dump file difference: call term_dumpdiff("Test_popup_command_01.dump.failed", "dumps/Test_popup_command_01.dump")
function RunTheTest[38]..Test_popup_command[17]..VerifyScreenDump line 14: See dump file difference: call term_dumpdiff("Test_popup_command_02.dump.failed", "dumps/Test_popup_command_02.dump")
TEST FAILURE
make[2]: *** [report] Error 1
make[1]: *** [scripttests] Error 2
make: *** [test] Error 2
```

**Comment by chrisbra** [comment]
can you run the mentioned term_dumpdiff() command?

**Comment by chdiza** [comment]
I can, and did.  But I don't understand what to do with it.  I can't compare the two "screens" unless I blow up my terminal to fullscreen.  When I do that, I can see no difference between the two except in the color of the highlighting.  Visible text is identical.

**Comment by nuko8** [comment]
The difference in color might give a hint as to what's going on with iTerm2 this time (Again, the failure took place only with iTerm2, not with Terminal.app, xterm, urxvt or st for me):

![screen shot 2018-03-06 at 22 14 31](https://user-images.githubusercontent.com/13826919/37039290-6952e270-219a-11e8-8e9d-aa351be72a2f.png)
```
call term_dumpdiff("Test_popup_command_01.dump.failed", "dumps/Test_popup_command_01.dump")
```
![screen shot 2018-03-06 at 22 15 07](https://user-images.githubusercontent.com/13826919/37039360-9e41cf0a-219a-11e8-9882-873abe29751b.png)
```
call term_dumpdiff("Test_popup_command_02.dump.failed", "dumps/Test_popup_command_02.dump")
```

**Comment by chdiza** [comment]
That's what I got too.

**Comment by chdiza** [comment]
I checked whether Terminal.app has the same failure before reporting this, and it didn't.  (I had run `make testclean`, switched to Terminal.app from iTerm2, cd'd to the build dir, and ran `make test` afresh and got the same failure.)  But then I reverted to a pristine working dir and tried building it fresh *in* Terminal.app and I got no test failure.  That seems weird.

**Comment by brammool** [comment]
Kazunobu Kuriyama wrote:

> The difference in color might give a hint as to what's going on with
> iTerm2 this time (Again, the failure took place only with iTerm2, not
> with Terminal.app, xterm, urxvt or st for me):
> 
> ![screen shot 2018-03-06 at 22 14 31](https://user-images.githubusercontent.com/13826919/37039290-6952e270-219a-11e8-8e9d-aa351be72a2f.png)
> ```
> call term_dumpdiff("Test_popup_command_01.dump.failed", "dumps/Test_popup_command_01.dump")
> ```
> ![screen shot 2018-03-06 at 22 15 07](https://user-images.githubusercontent.com/13826919/37039360-9e41cf0a-219a-11e8-9882-873abe29751b.png)
> ```
> call term_dumpdiff("Test_popup_command_02.dump.failed", "dumps/Test_popup_command_02.dump")
> ```

The background is a bit darker.  Does this terminal only support 16
colors?

-- 
BEDEVERE:        And what do you burn, apart from witches?
FOURTH VILLAGER: ... Wood?
BEDEVERE:        So why do witches burn?
SECOND VILLAGER: (pianissimo) ... Because they're made of wood...?
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chdiza** [comment]
> Does this terminal only support 16
colors?

No.  It supports TrueColor.

**Comment by nuko8** [comment]
The screenshots above were taken with iTerm2 of the factory default color settings.  According to Preferences...->Profiles->Default->Colors, the background color is #000000.  (FYI, the foreground color is #c7c7c7.)

According to the `src/testdir/Test_popup_command_{01,02}.dump.failed` which resulted from running the test with iTerm2 of that color settings, the background color of the popup is `#ff40ff` while the reference value is `#ffdfff`, and that of the selected item is `#6c6c6c` instead of `#e0e0e0`.  The rest of the colors are identical to the reference values, though.

In addition, on the `Colors` tab of the `Preferences...` popup, there's a pane for setting ANSI 16 colors.  The default values are:

Name | Normal | Brignt
------ |--------| ------
Black | `#000000` | `#686868`
Red | `#c91b00` | `#ff6e67`
Green | `#00c200` | `#5ffa68`
Yellow |`#c7c400`  | `#fffc67`
Blue | `#0225c7` | `#6871ff`
Magenta |`#ca30c7` | `#ff77ff`
Cyan | `#00c5c7` | `#60fdff`
White | `#c7c7c7` | `#ffffff`

So, it doesn't look so simple to conclude that somehow iTerm2 falls back on a 16 color palette against the test...

**Comment by chdiza** [comment]
I think the test failure was cured by patch 8.0.1615.  Anybody else notice this?

**Comment by chdiza** [comment]
Seems like I spoke too soon.  It only passes if `--disable-terminal` was passed.  I don't know why that would be; the test in question is testing vim's popup menu, IIUC.

**Comment by ichizok** [comment]
As far as I investigated; when testing `Test_popup_command ` on iTerm2 (with dark color window), the processings about 'background' option of vim on terminal window are:

1. set 'background' to 'light' (default value) at `set_init_1()`
2. set 'background' to 'dark' at `set_init_2()`; because iTerm2 has set `COLORFGBG` env variable, `term_bg_default()` returns "dark"
3. set 'background' to 'light' as the result of parsing the responses of t_RF/t_RB from libvterm

Actually popup is drawn between 2 and 3, thus the color of popup is based on 'background=dark'.

**Comment by chrisbra** [comment]
so simply setting $COLORFGBG to an empty string should work?

**Comment by nuko8** [comment]
> so simply setting $COLORFGBG to an empty string should work?

But that's unusual for iTerm2 and sounds like a hack only to pass the test.  I think we should consider doing away with use of $COLORFGBG.  After all, it's usually either 15;0 or 0;15, and has nothing to do with actual fg or bg color.

**Comment by ichizok** [comment]
Regarding Test_popup_command, I think those references (src/testdir/dumps/Test_popup_command_0x.dump) represent invalid state.

For example, the screen display of vim represented by Test_popup_comman_01.dump is below (upper left window):

![vim-test_popup v1](https://user-images.githubusercontent.com/943423/37731613-007b7c4c-2d86-11e8-9002-40ee9195b397.png)

However, this display is odd. There are the cursor and semicolon at 'showcmd' (or 'ruler') area.
I think it should be as below: the cursor is on 'X'.

![vim-test_popup v2](https://user-images.githubusercontent.com/943423/37732047-2eb0178e-2d87-11e8-8d1e-450e59cd23ef.png)

## Details

I ran the same processing as Test_popup_command_01 and logged it.

* set `ch_logfile()` for host vim (vim-A)
* enabled `DEBUG_TERMRESPONSE` and added logging at beginning of `check_termcode()` for on-terminal vim (vim-B)

chan.log
https://gist.github.com/ichizok/cd8ca2ef9a847bbacbe4ed6b1de5dc0f#file-chan-log
termresponse.log
https://gist.github.com/ichizok/cd8ca2ef9a847bbacbe4ed6b1de5dc0f#file-termresponse-log
Note: escape char (0x1b) is not shown on web browser.

Here, vim-A sent the response of t_RF/t_RB (and other) to vim-B,
https://gist.github.com/ichizok/cd8ca2ef9a847bbacbe4ed6b1de5dc0f#file-chan-log-L80

but vim-A received the part of it (`;rgb:ffff/`).
https://gist.github.com/ichizok/cd8ca2ef9a847bbacbe4ed6b1de5dc0f#file-chan-log-L109

That is, vim-B handled it as key-input, so caused the incorrect cursor and the extra semicolon.
Actually, `check_termcode()` of vim-B treated the separated response.
https://gist.github.com/ichizok/cd8ca2ef9a847bbacbe4ed6b1de5dc0f#file-termresponse-log-L385

The reason is because the vacancy of `typebuf.tb_buf` became not enough to store the response sequence owing to the input by `term_sendkeys()`.
Thus, for instance, we can confirm that when start vim-B with argument `+"set t_RF="` to shorten the response sequence there is the right display.

## Conclusion

I think the better way is adding `term_wait()` after `term_start()` (in `RunVimInTerminal()`) in order to wait for the completion of the initialization of vim-B.
In other words, I think it is desirable such that there is no need to mind the state of initialization of vim-B.

```diff
--- a/src/testdir/screendump.vim
+++ b/src/testdir/screendump.vim
@@ -41,6 +41,7 @@ func RunVimInTerminal(arguments, options)
   let cmd .= ' -v ' . a:arguments
   let buf = term_start(cmd, {'curwin': 1, 'term_rows': 20, 'term_cols': 75})
   call assert_equal([20, 75], term_getsize(buf))
+  call term_wait(buf, 200)

   return buf
 endfunc
```

**Comment by brammool** [comment]
Do you have a problem with test_syntax as well?  It checks the handling
of $COLORFGBG.

About this part:

> but vim-A received the part of it (`;rgb:ffff/`).
> https://gist.github.com/ichizok/cd8ca2ef9a847bbacbe4ed6b1de5dc0f#file-chan-log-L109

At first it is incomplete.  But later you see: 
  0.042719:    not enough characters for RB
  0.042755:    check_termcode: ']11;rgb:ffff/ffff/ffff'
  0.042763:    Received RBG response

So that works correctly.  The buffer only needs to be able to contain
the longest code completely, not multiple codes.

Also, the VerifyScreenDump() function waits for the screendump to match.
Thus if still waiting for the background color, it should cause a redraw
when it arrives and the correct screendump is found.

Perhaps a redraw is missing? Or the redraw is blocked while the popup
menu is showing.

Adding a sleep somewhere might help, but it's not a nice solution.
The sleep needs to happen before showing the popup.
The latest would be after sourcing menu.vim.  It would not have an
effect when doing it after ":popup PopUp".

-- 
Those who live by the sword get shot by those who don't.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ichizok** [comment]
> At first it is incomplete.  But later you see:
>   0.042719:    not enough characters for RB
>   0.042755:    check_termcode: ']11;rgb:ffff/ffff/ffff'
>   0.042763:    Received RBG response
> 
> So that works correctly.  The buffer only needs to be able to contain
> the longest code completely, not multiple codes.

Yes, finally parsing response was completed. But what I meant is that `;rgb:ffff/` should not be output into vim-A as a general.

> Also, the VerifyScreenDump() function waits for the screendump to match.
> Thus if still waiting for the background color, it should cause a redraw
> when it arrives and the correct screendump is found.
> 
> Perhaps a redraw is missing? Or the redraw is blocked while the popup
> menu is showing.

The subject of my comment is regarding to the display of status-line (cursor position and extra colon) rather than color.
Since it is the correct behavior of 'showcmd' in a sense,  it is probably not the matter about redrawing. Thus the wait in `VerifyScreenDump()` is useless for this problem.

> Adding a sleep somewhere might help, but it's not a nice solution.
> The sleep needs to happen before showing the popup.
> The latest would be after sourcing menu.vim.  It would not have an
> effect when doing it after ":popup PopUp".

`term_wait()` in `RunVimInTerminal()` is executed before any calling of `term_sendkeys()` so it has an effect.
But, as you said, simple wait is not good... maybe checking receiving contents is somewhat better. How is this?

```diff
--- a/src/testdir/screendump.vim
+++ b/src/testdir/screendump.vim
@@ -52,6 +52,7 @@ func RunVimInTerminal(arguments, options)
   let cmd .= ' -v ' . a:arguments
   let buf = term_start(cmd, {'curwin': 1, 'term_rows': 20, 'term_cols': 75})
   call assert_equal([20, 75], term_getsize(buf))
+  call WaitFor({-> !empty(filter(range(1, 20), {_, i -> term_getline(buf, i) !=# ''}))})

   return buf
 endfunc
```

**Comment by brammool** [comment]
Ozaki Kiichi wrote:

[...]

> > Also, the VerifyScreenDump() function waits for the screendump to match.
> > Thus if still waiting for the background color, it should cause a redraw
> > when it arrives and the correct screendump is found.
> > 
> > Perhaps a redraw is missing? Or the redraw is blocked while the popup
> > menu is showing.
> 
> The subject of my comment is regarding to the display of status-line
> (cursor position and extra colon) rather than color.
> Since it is the correct behavior of 'showcmd' in a sense,  it is
> probably not the matter about redrawing. Thus the wait in
> `VerifyScreenDump()` is useless for this problem.

What happens is that while the popup menu is displayed, other redraws
are postponed.  It's difficult to do otherwise, since the normal display
code doesn't know where the popup menu is, and the popup menu doesn't
know how to redraw other parts of the screen.  Would require a full
clear, redraw text and then redraw the popup menu, which no doubt cause
flicker.

I think I solved it for the test case by resetting $COLORFGBG before
running Vim.  I could reproduce the problem eventually and this made
it work.  I hope it also works in other circumstances.

> > Adding a sleep somewhere might help, but it's not a nice solution.
> > The sleep needs to happen before showing the popup.
> > The latest would be after sourcing menu.vim.  It would not have an
> > effect when doing it after ":popup PopUp".
> 
> `term_wait()` in `RunVimInTerminal()` is executed before any calling of `term_sendkeys()` so it has an effect.
> But, as you said, simple wait is not good... maybe checking receiving contents is somewhat better. How is this?
> 
> ```diff
> --- a/src/testdir/screendump.vim
> +++ b/src/testdir/screendump.vim
> @@ -52,6 +52,7 @@ func RunVimInTerminal(arguments, options)
>    let cmd .= ' -v ' . a:arguments
>    let buf = term_start(cmd, {'curwin': 1, 'term_rows': 20, 'term_cols': 75})
>    call assert_equal([20, 75], term_getsize(buf))
> +  call WaitFor({-> !empty(filter(range(1, 20), {_, i -> term_getline(buf, i) !=# ''}))})
> 
>    return buf
>  endfunc
> ```

Hmm, I think it's better for the caller to wait, the expected output may
depend on "arguments".

-- 
hundred-and-one symptoms of being an internet addict:
54. You start tilting your head sideways to smile. :-)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
